File: scripts/lib/taxing_bodies.py
"""Build a receipt's taxingBodies[] from the components the bill already prints.

Two rules shape this module.

**Roles are declared, never inferred.** The caller supplies a role for each
component, keyed by its sourceFactId. A label is not evidence of a role: the
receipt's own disclaimer says display names are never used to guess one, and
"Region of Waterloo" being an upper tier is a fact about Ontario's structure, not
about the string. An undeclared component is an error, so adding a fourth line to
a bill forces someone to say what it is rather than silently inheriting a
neighbour's role.

**Bodies are derived from the components, not written alongside them.** The
amounts a reader sees and the amounts in taxingBodies[] are then the same
numbers by construction. Maintaining two hand-written copies of a bill is how
they end up disagreeing, and a receipt whose breakdown contradicts its own total
is the worst thing this project could publish.

The invariants below are deliberately the same ones web/src/lib/taxingBodies.ts
re-checks in the browser. A reader should not have to trust bytes that travelled
over a network, and the builder should not be the only thing standing between a
malformed bill and a page.
"""

from __future__ import annotations

from decimal import Decimal
from typing import Any, Iterable, Mapping, Sequence

ROLES = ("local", "special-area", "upper-tier", "education")

# Roles a single bill may name at most once. Two local governments is a defect,
# not an unusual municipality.
SINGULAR_ROLES = ("local", "upper-tier", "education")


class TaxingBodyError(RuntimeError):
    """A bill could not be expressed as a coherent list of taxing bodies."""


def build_taxing_bodies(
    components: Sequence[Mapping[str, Any]],
    roles_by_fact_id: Mapping[str, str],
    *,
    total_cad: float,
    assessment_cad: int | None = None,
    basis: str,
    evidence_status: str = "DERIVED",
    ids_by_fact_id: Mapping[str, str] | None = None,
    notes_by_fact_id: Mapping[str, str] | None = None,
) -> list[dict[str, Any]]:
    """One body per component, in the order the bill prints them."""

    bodies: list[dict[str, Any]] = []
    for order, component in enumerate(components):
        fact_id = component.get("sourceFactId")
        if not fact_id:
            raise TaxingBodyError(
                f"component {order} ({component.get('label')!r}) has no sourceFactId, "
                "so its role cannot be declared against anything stable"
            )
        role = roles_by_fact_id.get(fact_id)
        if role is None:
            raise TaxingBodyError(
                f"no role declared for {fact_id!r} ({component.get('label')!r}). "
                "Add it to the role map; do not let it inherit a neighbour's role "
                "or be guessed from its label."
            )
        if role not in ROLES:
            raise TaxingBodyError(f"{fact_id!r} declares unknown role {role!r}; expected one of {ROLES}")

        body: dict[str, Any] = {
            "id": (ids_by_fact_id or {}).get(fact_id, fact_id.lower()),
            "role": role,
            "label": component["label"],
            "order": order,
            "amountCad": component["amountCad"],
            "basis": basis,
            "evidenceStatus": evidence_status,
            "sourceFactId": fact_id,
        }
        if assessment_cad is not None:
            body["assessmentCad"] = assessment_cad
        # Carry the rate across when the component declares one. A body's
        # amountCad is rate x assessment, but the body cited the rate fact while
        # printing the product and said nothing about the relationship - so the
        # printed number could not be checked against what it cited, and 12 of
        # these were unbound. With rate and assessmentCad both present the
        # arithmetic is stated by the artifact rather than inferred by a reader.
        rate = component.get("rate")
        # Decimal as well as float: these builders carry rates as Decimal so the
        # eight-place by-law rate is not rounded on the way in, and a plain
        # isinstance(int, float) check silently dropped every one of them.
        if isinstance(rate, Decimal):
            rate = float(rate)
        if isinstance(rate, (int, float)) and not isinstance(rate, bool):
            body["rate"] = rate
        note = (notes_by_fact_id or {}).get(fact_id)
        if note:
            body["note"] = note
        bodies.append(body)

    assert_bill_is_coherent(bodies, total_cad)
    return bodies


def assert_bill_is_coherent(
    bodies: Iterable[Mapping[str, Any]],
    total_cad: float | None,
    inapplicable: Iterable[Mapping[str, Any]] = (),
) -> None:
    """Refuse a bill that cannot be true, rather than warning about it."""

    bodies = list(bodies)
    counts: dict[str, int] = {}
    for body in bodies:
        counts[body["role"]] = counts.get(body["role"], 0) + 1

    for role in SINGULAR_ROLES:
        if counts.get(role, 0) > 1:
            raise TaxingBodyError(
                f"a bill may name at most one {role} body; this one names {counts[role]}"
            )
    if counts.get("local", 0) == 0:
        raise TaxingBodyError("a bill must name the local municipality that issues it")

    charged = {body["role"] for body in bodies}
    for entry in inapplicable:
        if entry["role"] in charged:
            raise TaxingBodyError(
                f"{entry['role']} is listed both as a taxing body and as not applicable"
            )

    if total_cad is not None:
        summed = round(sum(body["amountCad"] for body in bodies), 2)
        # A cent of rounding across four bodies is expected; a dollar is not.
        if abs(summed - total_cad) > 0.05:
            raise TaxingBodyError(
                f"taxing bodies sum to {summed:.2f} but the receipt prints {total_cad:.2f}"
            )

File: web/src/lib/taxingBodies.ts
/**
 * Read the bill as a list of taxing bodies.
 *
 * Roles are declared by the builder and never inferred here. That is not
 * caution for its own sake: the existing artifacts make inference impossible to
 * do honestly. The three legacy buckets and combinedAtAssessment.components are
 * different measurements - North Dumfries' region bucket is $2,543 at a $354,500
 * assessment while its region component is $3,264.83 at $455,000 - and their
 * sourceFactIds do not join, because a bucket cites an allocation fact and a
 * component cites a rate fact. The only remaining way to pair them would be to
 * match on the display label, and the receipt's own disclaimer says display
 * names are never used to guess a role.
 *
 * So a receipt that has not declared its bodies is refused rather than guessed
 * at. The error names the builder, because that is where the fix belongs.
 *
 * What this unlocks, visible in the packs already committed:
 *
 *   County of Brant is single-tier and today carries a placeholder bucket
 *   labelled "Upper-tier (n/a)" because the field is required. A shorter bill
 *   should be a shorter list.
 *
 *   Brant residents also pay a $78.04 hospital special levy. It is its own line
 *   in combinedAtAssessment, and with only three slots it is currently folded
 *   into the municipal portion, so the receipt understates what the county
 *   charges and hides a levy the reader is entitled to see.
 */

import type {
  InapplicableBody,
  TaxingBody,
  TaxingBodyRole,
  TaxpayerReceipt,
} from '../types'

export class TaxingBodyError extends Error {}

type Profile = TaxpayerReceipt['profiles']['supportedAverageHousehold']

export type Bill = {
  bodies: TaxingBody[]
  inapplicable: InapplicableBody[]
}

/** Bill order. A reader expects local money first and the province last. */
const ROLE_ORDER: TaxingBodyRole[] = ['local', 'special-area', 'upper-tier', 'education']

/** Roles that may appear at most once. Two local governments is a defect. */
const SINGULAR_ROLES: TaxingBodyRole[] = ['local', 'upper-tier', 'education']

/**
 * Check what a bill must be true of before anything renders it.
 *
 * Deliberately not "warn": a receipt whose parts do not add to its printed
 * total is the single worst thing this project could publish, and it is
 * cheaper to refuse than to explain afterwards.
 */
export function assertBillIsCoherent(bill: Bill, combinedTotalCad: number | null): void {
  const counts = new Map<TaxingBodyRole, number>()
  for (const body of bill.bodies) {
    counts.set(body.role, (counts.get(body.role) ?? 0) + 1)
  }
  for (const role of SINGULAR_ROLES) {
    const n = counts.get(role) ?? 0
    if (n > 1) {
      throw new TaxingBodyError(`A bill may name at most one ${role} body; this one names ${n}.`)
    }
  }
  if ((counts.get('local') ?? 0) === 0) {
    throw new TaxingBodyError('A bill must name the local municipality that issues it.')
  }
  const overlap = bill.inapplicable.filter((entry) =>
    bill.bodies.some((body) => body.role === entry.role),
  )
  if (overlap.length > 0) {
    throw new TaxingBodyError(
      `${overlap[0].role} is listed both as a taxing body and as not applicable.`,
    )
  }
  if (combinedTotalCad !== null) {
    const summed = bill.bodies.reduce((sum, body) => sum + body.amountCad, 0)
    // A cent of rounding across four bodies is expected; a dollar is not.
    if (Math.abs(summed - combinedTotalCad) > 0.05) {
      throw new TaxingBodyError(
        `Taxing bodies sum to ${summed.toFixed(2)} but the receipt prints ${combinedTotalCad.toFixed(2)}.`,
      )
    }
  }
}

/** The bill for a receipt, in reading order. Declared, never inferred. */
export function taxingBodiesFor(profile: Profile, jurisdictionSlug?: string): Bill {
  const declared = profile.taxingBodies
  if (!declared || declared.length === 0) {
    throw new TaxingBodyError(
      `${jurisdictionSlug ?? 'This receipt'} does not declare taxingBodies[]. ` +
        'Roles cannot be recovered from the legacy buckets: they measure something ' +
        'else, at a different assessment, and their fact ids do not join. The ' +
        'builder that produced this artifact has to emit the bodies and their roles.',
    )
  }

  const bill: Bill = {
    bodies: [...declared].sort(
      (a, b) =>
        a.order - b.order || ROLE_ORDER.indexOf(a.role) - ROLE_ORDER.indexOf(b.role),
    ),
    inapplicable: profile.inapplicableBodies ?? [],
  }
  assertBillIsCoherent(bill, profile.combinedTotalCad ?? null)
  return bill
}

/**
 * The bill if this receipt declares one, and null if it does not.
 *
 * taxingBodiesFor() refuses an undeclared receipt, which is right for anything
 * that needs the roles: guessing them is the failure this module exists to
 * prevent. But a screen has to render the five packs whose builders have not
 * been migrated yet, and "no declaration" is a different condition from "a
 * declaration that does not hold". This returns null for the first and still
 * throws for the second, so a migrated pack can never quietly fall back to the
 * legacy path when its bill is wrong.
 */
export function declaredBillFor(profile: Profile, jurisdictionSlug?: string): Bill | null {
  if (!profile.taxingBodies || profile.taxingBodies.length === 0) return null
  return taxingBodiesFor(profile, jurisdictionSlug)
}

/** Each body's share of the bill, for the bar and the legend. */
export function billShares(bill: Bill): { body: TaxingBody; share: number }[] {
  const total = bill.bodies.reduce((sum, body) => sum + body.amountCad, 0)
  if (total <= 0) return bill.bodies.map((body) => ({ body, share: 0 }))
  return bill.bodies.map((body) => ({ body, share: body.amountCad / total }))
}
